File: javascript/tree/543_Diameter_of_Binary_Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {
    let res = 0;

    dfs = (root) => {
        if (!root) return -1;

        const left = dfs(root.left);
        const right = dfs(root.right);

        res = Math.max(res, left + right + 2);

        return 1 + Math.max(left, right);
    }
    dfs(root);
    return res;
};

File: python/tree/543_Diameter_of_Binary_Tree.py
'''
brute force: Time: O(n^2)
bottom up:
Time: O(n)
'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 0  # remove nonlocal, let res = [0]

        def dfs(root):
            nonlocal res  # no nonlocal: Unbound error  
            if not root:
                return -1  # height of null node = -1

            left = dfs(root.left)
            right = dfs(root.right)
            res = max(res, left + right + 2)  # diameter = (left height + 1) + (right height + 1)
            
            return 1 + max(left, right)  # current height = 1 + max(left height, right height)

        dfs(root)
        return res
